Add ability to migrate Alerting Config to migrate_rules

aiyengar2 opened the pull request:
Related Issue: https://github.com/rancher/rancher/issues/29235

Commit:
Add ability to migrate Alerting config to migrate_rules

aiyengar2 reviewed (commented):
@bashofmann the approach taken in this PR mostly takes the user's existing Alertmanager YAML and modifies it instead of trying to create the Alertmanager YAML from scratch. This has the following caveats:
- It won't catch notifiers declared by the user in Monitoring V1 with no routes enabled for them (since those wouldn't be in the Monitoring V1 Alertmanager YAML until at least one route has been added for it)
- Webhook Config based Notifiers (DingTalk, Microsoft Teams) will be broken the moment the webhook-receiver deployment in the cattle-prometheus namespace is removed (since the URL would need to be modified after the user redeploys this in Rancher 2.5 for Monitoring V2).
- I completely discarded rule-specific `group_interval`, `group_wait`, and `repeat_interval` values provided by the user (see `del route["routes"]`) since routes with child routes isn't natively supported in the Monitoring V2 UI.

Do these caveats sound okay to you?

bashofmann commented on migrate_rules.py:
```diff
@@ -64,62 +64,196 @@ def migrate(rancher_url, rancher_api_token, cluster_id, project_id, insecure):
+    alerting_config_response = requests.get(alerting_config_url, headers=headers, verify=verify)
+    alerting_config = json.loads(alerting_config_response.content)
+    alertmanager_yaml = yaml.safe_load(
+        base64_decode(alerting_config["data"]["alertmanager.yaml"])
```
This fails if the cluster has not alertmanager config (`KeyError: 'data'`). I think we should handle this more gracefully.

bashofmann replied:
Response is `{'baseType': 'error', 'code': 'NotFound', 'message': 'secrets "alertmanager-cluster-alerting" not found', 'status': 404, 'type': 'error'}`

aiyengar2 replied:
Good catch. Seems like it would be a quick fix to just check when `alerting_config_response.status_code == 404` before trying to load the `alertmanager_yaml` here.

If we can't find the Alertmanager Secret, I think the behavior of this script should be to only output PrometheusRules and drop the Namespace / Secret resource in favor on Monitoring V2 creating it for the user.

aiyengar2 replied:
Has been updated for this. Also prints out nice log statements to stderr.

bashofmann commented on migrate_rules.py:
```diff
@@ -64,62 +64,196 @@ def migrate(rancher_url, rancher_api_token, cluster_id, project_id, insecure):
+    alertmanager_yaml["route"]["receiver"] = "null"
+    alertmanager_yaml["route"]["group_by"] = ['job']
+    routes = []
+    for route in alertmanager_yaml["route"]["routes"]:
```
I think the generated config should still contain the default route to explicitely send the Watchdog alert to the null receiver:
```
      routes:
      - match:
          alertname: Watchdog
        receiver: 'null'
```

aiyengar2 replied:
Sounds good, will add it to the PR.

aiyengar2 commented on migrate_rules.py:
```diff
@@ -64,62 +64,196 @@ def migrate(rancher_url, rancher_api_token, cluster_id, project_id, insecure):
+
+    alertmanager_yaml = update_alertmanager_config(alertmanager_yaml, notifiers_by_id)
+
+    print(yaml.dump(alertmanager_yaml), end="---\n")
```
```suggestion
```

Commit:
Fail gracefully if Alerting V1 is disabled

- Print warning to stderr if the Alertmanager Config Secret cannot be extracted or if no resources can be ported over
- Add checks throughout the script to support using it even if Alerting V1 is disabled to port over PrometheusRules
- Misc: add Watchdog route by default

aiyengar2 commented:
If you have no Alertmanager Secret but valid metric rule that can be ported over:
```
$ python3 migrate_rules.py --rancher-url=<redacted> --rancher-api-token=<redacted> --cluster-id=c-ms7nm > output.yaml

Warning: Cluster Alerting seems to be disabled.

Could not extract any Alertmanager Config for this cluster.

Any metric-based Alerting Groups / Alerting Rules have been outputted as PrometheusRule resources.

However, you will need to manually configure Routes and Receivers to set up notifications based on those alerts.

See Rancher docs for more information on how to configure notifications on alerts in Monitoring V2.

```

If you have no valid metric rules:
```
$ python3 migrate_rules.py --rancher-url=<redacted> --rancher-api-token=<redacted> --cluster-id=c-ms7nm > output.yaml

Warning: Cluster Alerting seems to be disabled.

Could not extract any Alertmanager Config or any PrometheusRule resources from this cluster.

If you believe resources should have been picked up for migration, please file a bug or feature request with Rancher at https://github.com/rancher/rancher/issues/new.

```
